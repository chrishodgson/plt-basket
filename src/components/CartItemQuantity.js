import React, { Component } from 'react';
import { connect } from "react-redux";
import { increaseQuantity, decreaseQuantity, removeProduct } from "../actions/index";

class CartItemQuantity extends Component {

    state = { quantity: 0 };

    onIncreaseQuantity() {
        this.setState(prevState => {
            this.props.increaseQuantity(this.props.product.id);
            return { quantity: prevState['quantity'] ? prevState['quantity'] + 1 : 1 }
        });
    }
    onDecreaseQuantity() {
        this.setState(prevState => {
            this.props.decreaseQuantity(this.props.product.id);
            return { quantity: prevState['quantity'] ? prevState['quantity'] - 1 : 0 }
        });
    }
    onRemoveProduct() {
        this.setState(() => {
            this.props.removeProduct(this.props.product.id);
            return { quantity: 0 }
        });
    }

    render() {
        return (
            <div className="cartItemQuantity">
                <button className="btn btn-secondary btn-xs" onClick={this.onDecreaseQuantity.bind(this)}>-</button>
                <span>{this.state.quantity}</span>
                <button className="btn btn-secondary btn-xs" onClick={this.onIncreaseQuantity.bind(this)}>+</button>
                <p><button className="btn btn-link" onClick={this.onRemoveProduct.bind(this)}>remove</button></p>
            </div>
        );
    }
}

export default connect(
    ({ products }) => ({ products }),
    { increaseQuantity, decreaseQuantity, removeProduct }
)(CartItemQuantity);
