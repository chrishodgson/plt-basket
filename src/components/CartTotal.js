import React, { Component } from 'react';
import { connect } from "react-redux";
import { productColourSelector } from '../selectors/index'

class CartTotal extends Component {

    getCartTotal() {
        let total = 0;
        for (let product of productColourSelector(this.props.colour, this.props.products)) {
            const quantity = product['quantity'] || 0;
            if (quantity) {
                total += quantity * product['price'];
            }
        }
        return total.toFixed(2);
    }

    render() {
        return (
            <h3 className="cartTotal">
                &pound;{this.getCartTotal()}
            </h3>
        );
    }
}

function mapStateToProps({ form, products }) {
    const colour = form.colourFilter && form.colourFilter.values && form.colourFilter.values.colour || '';
    return { colour, products };
}

export default connect(mapStateToProps)(CartTotal);
