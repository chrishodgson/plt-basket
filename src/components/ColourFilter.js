import React, {Component} from 'react';
import { reduxForm, Field } from "redux-form";
import { connect } from "react-redux";

class ColourFilter extends Component {

    getColours() {
        const colours = [];
        for (let product of this.props.products) {
            if (!colours.includes(product['colour'])) {
                colours.push(product['colour']);
            }
        }
        return colours;
    }

    renderFields() {
        return (
                <option value="all">-- Choose a colour --</option>
                {this.getColours().map(colour => <option key={colour} value={colour}>{colour}</option>)}
            </Field>
        );
    }

    render() {
        return (
            <form onSubmit={this.props.handleSubmit}>
                {this.renderFields()}
            </form>
        );
    }
}

export default reduxForm({
    form: "colourFilter"
})(
    connect( ({ products }) => ({ products }) )(ColourFilter)
);
