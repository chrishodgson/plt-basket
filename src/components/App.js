import React, { Component } from 'react';

import ColourFilter from './ColourFilter';
import ProductListing from './ProductListing';
import CartTotal from './CartTotal';

export default class App extends Component {
  render() {
    return (
        <div>
            <ColourFilter/>
            <ProductListing/>
            <CartTotal/>
        </div>
    );
  }
}