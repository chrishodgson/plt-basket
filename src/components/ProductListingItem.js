import React from 'react';
import { create } from "react-test-renderer";
import { productListingItem } from '../../src/components/ProductListingItem'

export default ( {product: {img, name, price}} ) => {
    return (
        <div className="productListingItem">
            <img src={img} alt={name} />
            <div className="productListingItemText">
                <div>{name}</div>
                <div>&pound;{price.toFixed(2)}</div>
            </div>
        </div>
    );
};
