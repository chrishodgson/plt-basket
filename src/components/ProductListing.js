import React, { Component } from 'react';
import { connect } from "react-redux";
import { fetchProducts } from "../actions/index";
import { productColourSelector } from '../selectors/index'

import ProductListingItem from './ProductListingItem';
import CartItemQuantity from './CartItemQuantity';

class ProductListing extends Component {

    componentDidMount() {
        this.props.fetchProducts();
    }

    renderProducts() {
        return productColourSelector(this.props.colour, this.props.products).map(product => {
            return (
                <li className="list-group-item" key={product.id}>
                    <ProductListingItem product={product} />
                    <CartItemQuantity product={product} />
                </li>
            );
        });
    }

    render() {
        return (
            <ul className="list-group">
                {this.renderProducts()}
            </ul>
        );
    }
}

function mapStateToProps({ form, products }) {
    const colour = form.colourFilter && form.colourFilter.values && form.colourFilter.values.colour || '';
    return { colour, products };
}

export default connect(mapStateToProps, { fetchProducts })(ProductListing);
