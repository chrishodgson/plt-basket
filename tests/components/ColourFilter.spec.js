import React from "react";
import { create } from "react-test-renderer";
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { ColourFilter } from "../../src/components/ColourFilter";

jest.mock('redux-form/lib/Field', () => 'field');

Enzyme.configure({ adapter: new Adapter() });

describe("ColourFilter component", () => {

    const products = [ {colour:'red'}, {colour:'yellow'}, {colour:'red'} ];

    const shallowSetup = () => {
        const props = { handleSubmit: jest.fn(), products };

        const enzymeWrapper = shallow(<ColourFilter {...props} />);

        return { props, enzymeWrapper };
    };

    test("Matches the snapshot", () => {
        const tree = create(<ColourFilter products={products}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should fire handleSubmit', () => {
        const { props, enzymeWrapper } = shallowSetup(),
            form = enzymeWrapper.find('form').first();
        form.simulate('submit');
        expect(props.handleSubmit.mock.calls.length).toBe(1);
    });

    it('should render correct colour options', () => {
        const { enzymeWrapper } = shallowSetup(),
            options = enzymeWrapper.find('option');

        expect(options.length).toBe(3);
        expect(options.first().prop('value')).toBe('all');
        expect(options.at(1).prop('value')).toBe('red');
        expect(options.last().prop('value')).toBe('yellow');
    });
});
