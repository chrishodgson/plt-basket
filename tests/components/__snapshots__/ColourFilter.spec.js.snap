// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColourFilter component Matches the snapshot 1`] = `
<form>
  <field
    className="colourFilter"
    component="select"
    name="colour"
  >
    <option
      value="all"
    >
      -- Choose a colour --
    </option>
    <option
      value="red"
    >
      red
    </option>
    <option
      value="yellow"
    >
      yellow
    </option>
  </field>
</form>
`;
